Plugins/SequencePlugin/Source/SequencePlugin/Private/ResponseSignatureValidator.cpp

```diff
@@ -0,0 +1,195 @@
+    {
+        // Extract the key values
+        const TArray<TSharedPtr<FJsonValue>> Keys = JsonObject->GetArrayField("keys");
+        const TSharedPtr<FJsonObject> Key = Keys[0]->AsObject();
+
+        FString ExponentStr = Key->GetStringField("e");
+        FString ModulusStr = Key->GetStringField("n");
+
+        // Decode the Base64 URL values
+        TArray<uint8> ExponentBytes = Base64UrlDecode(ExponentStr);
+        TArray<uint8> ModulusBytes = Base64UrlDecode(ModulusStr);
+
+        // Convert to OpenSSL BIGNUM format
+        BIGNUM* ExponentBN = BN_bin2bn(ExponentBytes.GetData(), ExponentBytes.Num(), NULL);
+        BIGNUM* ModulusBN = BN_bin2bn(ModulusBytes.GetData(), ModulusBytes.Num(), NULL);
+
+        // Create the RSA key structure
+        WaasPublicKey = RSA_new();
+        RSA_set0_key(WaasPublicKey, ModulusBN, ExponentBN, NULL);
+
+        PublicKeyFetched = true;
+    }
+    else
+    {
+        UE_LOG(LogTemp, Error, TEXT("Error loading public key from JWKS"));
+        PublicKeyFetched = false;
+    }
+}
+
+TArray<uint8> UResponseSignatureValidator::Base64UrlDecode(const FString& Input)
+{
+    FString Output = Input;
+    Output = Output.Replace(TEXT("-"), TEXT("+")).Replace(TEXT("_"), TEXT("/"));
+
+    // Adjust for padding
+    while (Output.Len() % 4 != 0)
+    {
+        Output.AppendChar('=');
+    }
+
+    TArray<uint8> DecodedBytes;
+    FBase64::Decode(Output, DecodedBytes);
+    return DecodedBytes;
+}
+
+FString UResponseSignatureValidator::GetValue(const FString& HeaderItem)
+{
+    int32 FirstColon, LastColon;
+    if (HeaderItem.FindChar(':', FirstColon) && HeaderItem.FindLastChar(':', LastColon) && FirstColon != LastColon)
+    {
+        return HeaderItem.Mid(FirstColon + 1, LastColon - FirstColon - 1).TrimStartAndEnd();
+    }
+    return FString(); // Return empty if no valid value found.
+}
+
+bool UResponseSignatureValidator::ValidateResponseSignature(const FHttpResponsePtr& Response)
+{
+    if (!Response.IsValid())
+        return false; 
+
+    const FString ContentDigestHeader = Response->GetHeader(TEXT("content-digest"));
+    const FString SignatureHeader = Response->GetHeader(TEXT("signature"));
+    const FString SignatureInputHeader = Response->GetHeader(TEXT("signature-input"));
+
+    if (ContentDigestHeader.IsEmpty() || SignatureHeader.IsEmpty() || SignatureInputHeader.IsEmpty())
+        return false;
+
+    TArray<uint8> ResponseContent = Response->GetContent();
+    FString CleanContentDigestHeader = ContentDigestHeader;
+    const FString Prefix = TEXT("sha-256=");
+    if (CleanContentDigestHeader.StartsWith(Prefix))
+        CleanContentDigestHeader = CleanContentDigestHeader.RightChop(Prefix.Len());
+
+    CleanContentDigestHeader.RemoveAt(0);
+    CleanContentDigestHeader.RemoveAt(CleanContentDigestHeader.Len() - 1);
+
+    TArray<uint8_t> DecodedDigest;
+
+    if (!FBase64::Decode(CleanContentDigestHeader, DecodedDigest))
+        return false;   
```

Comment:
Similar to the above, let's log an error message if this were to happen to make error diagnosis easier (even though this is rather unlikely to fail)

Reply:
done